### ISSUE TITLE
tmux writing extraneous lines to scrollback on line 1 CR+CSI ED

### ISSUE THREAD

**Issue by robolange** [tmux/tmux#3090]
### Issue description

I've noticed that under certain circumstances, tmux pushes extraneous lines to its scrollback buffer. After hacking around for a bit on it, I think I've developed a minimal reproducer that shows the ANSI sequence that causes this.

In short, if a sequence like `foo^M^[[Jbar` is emitted in the first line of the tmux pane, tmux will push a line `foo` into the scrollback. I've tested this in several terminals outside of tmux (Apple Terminal, Gnome Terminal, rxvt, linux console) and none of those terminals pushed such a line to scrollback when given that input.

Based on my (admittedly limited) knowledge of ANSI sequences, I was surprised that a CR + CSI ED would cause a line to be pushed to scrollback. I would expect it simply to move to the beginning of the current line and erase the screen from that point forward, as it does in all of the terminals I tested outside of tmux. Furthermore, this only occurs in tmux when emitted on the first line. When emitted on a subsequent line, the behavior is as expected.

I'm not sure if this is a regression, or if tmux has always done this. The oldest tmux version I have handy is 3.1, and this happens on that version too.

Update: I've tested against the latest master branch and can reproduce it there as well.

A minimal reproducer, `test.zsh`:
```
#!/usr/bin/env zsh
#
# In tmux, this incorrectly(?) pushes `test aaa` to scrollback. Outside of
# tmux, nothing is pushed to scrollback (on gnome-terminal, rxvt, or
# Apple Terminal).

clear
[ -n "${TMUX-}" ] && tmux clearhist
printf '%s\n' $'\e[Jtest aaa\CM\e[Jtest bbb'
```
If you don't have zsh, you can make an equivalent bash script by changing `\C` in the printf statement to `\c`. The exhibited behavior is the same.

### Required information

Apple Terminal, outside of tmux:
```
% uname -sp && tmux -V && echo $TERM
Darwin i386
tmux 3.2a
xterm-256color
```
Also reproduced on Gnome Terminal (TERM=xterm-256color), rxvt (TERM=rxvt-unicode-256color), and Linux console (TERM=linux).

No tmux configuration is required to reproduce. Inside tmux (on all terminals):
```
% echo $TERM
screen
```
Please see attached tmux logs.
[tmux-logs.tar.gz](https://github.com/tmux/tmux/files/8149115/tmux-logs.tar.gz)

**Comment by nicm** [comment]
ED0 from the first position is the same as ED2 and will clear into the history.

This is intentional:
~~~~
        /* Scroll into history if it is enabled and clearing entire screen. */
        if (s->cx == 0 && s->cy == 0 && (gd->flags & GRID_HISTORY))
                grid_view_clear_history(gd, bg);
        else {
~~~~
Now *why* is the question... we definitely want to do it on ED0, but I don't remember if ED2 was included for completeness or for any particular reason.

**Comment by nicm** [comment]
Oh your example is using ED0 actually not ED2, so this seems to be working as intended?

**Comment by nicm** [comment]
This behaviour was added in ecc22c521d83f8ce480c0e2866fdf3ed14bd6131 but I don't remember anything else about it.

**Comment by nicm** [comment]
xterm can do this, but it is turned off by default:
~~~~
       cdXtraScroll (class CdXtraScroll)
               Specifies whether xterm should scroll to a new page when
               clearing the whole screen.  Like tiXtraScroll, the intent of
               this option is to provide a picture of the full-screen
               application's display on the scrollback before wiping out the
               text.

               Xterm accepts either a keyword (ignoring case) or the number
               shown in parentheses:

               false (0)
                      nothing is added to the scrollback.

               true (1)
                      the current screen is added to the scrollback.

               trim (2)
                      the current screen is added to the scrollback, but
                      repeated blank lines are trimmed (reduced to a single
                      blank line).

               The default for this resource is “false”.
~~~~

**Comment by robolange** [comment]
Thanks for the quick followup. Would you be open to a patch adding an option for this behavior to tmux?

**Comment by robolange** [comment]
Zsh does a `^M^[[J` when redrawing the prompt. It does an `^[[J` instead of an `^[[K` because  because it doesn't just redraw that line, it also needs to redraw multi-line commands and any completions that are on-screen. Redraws happen for a variety of reasons, including SIGWINCH, which happens any time I spit a tmux pane.

So what'll happen is I'll be writing a command, using completions to help, then I'll split open a new pane to read a man page, kill the split when I'm done reading, go back to editing the command. And then I realize that my scrollback is polluted with a bunch of half-written copies of the command and whatever zsh completions were on-screen when I did those splits. This probably wouldn't be so bad, but I'm also in the habit of doing a `clear` between logically-related sets of commands, so I'm fairly often editing a prompt that's on line 1.

I used the same workflow with tmux+bash for years, but have recently switched to zsh, which exposed this behavior.
